lib/vrl/compiler/src/type_def.rs

```diff
@@ -951,6 +1150,37 @@ impl TypeDef {
             kind: self.kind.merge(rhs.kind, false, true),
         }
     }
+
+    /// For any array type defs within this type, allows you to map the indexed type def to a new
+    /// type def.
+    pub fn map_array<F>(&self, f: F) -> Self
+    where
+        F: Fn(TypeDef) -> TypeDef,
+    {
+        let newkind = self.kind.map(|k| match k {
+            TypeKind::Array(array) => TypeKind::Array(
+                array
+                    .iter()
+                    .map(|(index, kind)| {
+                        (
+                            *index,
+                            f(TypeDef {
+                                fallible: self.fallible,
+                                kind: kind.clone(),
+                            })
+                            .kind,
```

Comment:
I'm not completely sold on this. In an attempt to avoid exposing `KindInfo` as public it is being wrapped in `TypeDef`, passed to the closure, and `KindInfo` is then unwrapped (any changes made to `fallible` in are thrown away).

The alternative - making `KindInfo` public does lead to a cascade of other types needing to be exposed.

I'm not sure which is the best solution here.

Reply:
Why wouldn't you update `fallible` as well?

Reply:
I don't think it really makes sense to. 

The array may contain multiple rows, and so the closure would be called multiple times - the order of which is largely arbitrary. Each row is actually a `TypeKind` that is only being wrapped in a `TypeDef` for the purposes of this call. 

Ultimately it would be the last call that would determine the value of the resulting `TypeDef` fallible.